### PR TITLE
Global Pooling Collapse Dimension change

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Global pooling collapse dimensions returned input type feed forward.
This sometimes made interacting with it inconsistent.
Since GlobalPooling is mainly used in the RNN context, it's preferrable to add a pre processor
where necessary and just keep the rnn type for most interactions rather than collapsing to feed forward.

(Please fill in changes proposed in this fix)
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

**Comment by joel-a** [comment]
@agibsonccc I checked out the fix branch and build new jars from it. Depending on how I run my test, I'm either seeing the previous error or the following:
```
Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Arrays are different shape and are not broadcastable. Array 1 shape = [1200, 512], array 2 shape = [3, 400, 512]
	at org.nd4j.linalg.api.shape.Shape.assertBroadcastable(Shape.java:3568)
	at org.nd4j.linalg.api.shape.Shape.assertBroadcastable(Shape.java:3561)
	at org.nd4j.linalg.api.shape.Shape.broadcastOutputShape(Shape.java:291)
	at org.deeplearning4j.nn.graph.vertex.impl.ElementWiseVertex.doForward(ElementWiseVertex.java:98)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputOfLayersDetached(ComputationGraph.java:2438)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1740)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1696)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1626)
```
I hope that helps in finding a fix and thank you for your prompt attention to this issue.

**Comment by agibsonccc** [comment]
@joel-a are you running it the same? I was running it according to your input before for the initial test. Either way thanks for the feedback!

Edit: Make sure you install both deeplearning4j-nn and deeplearning4j-modelimport. Whenever a model import fix is applied, sometimes it involves changing some of the assumptions in the core layers themselves.

**Comment by joel-a** [comment]
@agibsonccc I'm running the same model and yes, I'm installing both nn and modelimport. I realized the issue is during inference not while loading. The problem seems to be the either the RepeatVector layer or the Activation layer is outputting a tensor of the wrong dimensions. The first two dimensions of what's supposed to be a 3d tensor, are being multiplied to create the first dimension of a 2d tensor. When the multiplication layer tries to multiply both outputs, RepeatLayer and Activation, the tensor shapes don't match.  (See error message below)

**Error:**
```
Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Arrays are different shape and are not broadcastable. Array 1 shape = [1200, 512], array 2 shape = [3, 400, 512]
	at org.nd4j.linalg.api.shape.Shape.assertBroadcastable(Shape.java:3568)
	at org.nd4j.linalg.api.shape.Shape.assertBroadcastable(Shape.java:3561)
	at org.nd4j.linalg.api.shape.Shape.broadcastOutputShape(Shape.java:291)
	at org.deeplearning4j.nn.graph.vertex.impl.ElementWiseVertex.doForward(ElementWiseVertex.java:98)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputOfLayersDetached(ComputationGraph.java:2438)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1740)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1696)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1626)
```

Again, here's the code to generate the model which is also attached as a zip file for convenience (no changes from previous post):
```
#Model to replicate error
in_layer = Input(shape=(400, 128))
x = Conv1D(512, 3, strides=1, dilation_rate=1, padding='same', activation='linear')(in_layer)
x = BatchNormalization()(x)
mask = Activation('relu')(x)
#out_layer = Conv1D(512, 3, strides=1,
#                          dilation_rate=3,
#                          padding='same',
#                          activation='linear')(x)
se_mask = GlobalAveragePooling1D()(mask)
se_mask = Dense(int(512 / 10), activation='relu')(se_mask)
se_mask = Dense(512, activation='sigmoid')(se_mask)
se_mask = RepeatVector(400)(se_mask)
out_layer = multiply([mask, se_mask])
gapModel = Model(in_layer, out_layer)
gapModel.save("GAPError.h5")
```

**To load and run in DL4j**
```
String modelPath = "GAPError.h5";
ComputationGraph model = KerasModelImport.importKerasModelAndWeights(modelPath);
String inputPath = "encoderOutput.npy";
INDArray input = Nd4j.createFromNpyFile(new File(inputPath));
        
//Run inference
var output = model.output(input)[0];
```
Thank you!

[GAPError.zip](https://github.com/eclipse/deeplearning4j/files/6716674/GAPError.zip)
[encoderOutput.zip](https://github.com/eclipse/deeplearning4j/files/6716737/encoderOutput.zip)

**Comment by agibsonccc** [comment]
@joel-a yeah I was able to run inference just fine with that exact model. I'm not quite sure what the difference would be.  Only thing I can think of is I'm using tf keras to save the model rather than normal keras. What version of keras are you using? Either way, inference for me runs just fine.
[GAPERROR_run.zip](https://github.com/eclipse/deeplearning4j/files/6716865/GAPERROR_run.zip)

Here's the converted model I ran. 

Here's the simple main output I ran:
```
o.d.n.m.k.KerasModel - If enforceTrainingConfig is true, a training configuration object has to be provided. Usually the only practical way to do this is to store your keras model with `model.save('model_path.h5')`. If you store model config and weights separately no training configuration is attached.
o.n.l.f.Nd4jBackend - Loaded [CpuBackend] backend
o.n.n.NativeOpsHolder - Number of threads used for linear algebra: 8
o.n.l.c.n.CpuNDArrayFactory - Binary level Generic x86 optimization level AVX/AVX2
o.n.n.Nd4jBlas - Number of threads used for OpenMP BLAS: 8
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CPU]; OS: [Windows 10]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [16]; Memory: [7.1GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
o.n.l.c.n.CpuBackend - Backend build information:
 GCC: "10.2.0"
STD version: 201103L
DEFAULT_ENGINE: samediff::ENGINE_CPU
HAVE_FLATBUFFERS
HAVE_OPENBLAS
o.d.n.g.ComputationGraph - Starting ComputationGraph with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
Rank: 3, DataType: FLOAT, Offset: 0, Order: c, Shape: [1,400,512],  Stride: [204800,512,1]
```

Here's the java code:
```java
  ComputationGraph computationGraph = KerasModelImport.importKerasModelAndWeights("GAPError.h5");
        INDArray input = Nd4j.ones(1,400,128);
        INDArray[] output = computationGraph.output(input);
        System.out.println(output[0].shapeInfoToString());
        computationGraph.save(new File("GAPERROR_run.zip"));
```

Edit: Just to make sure, I do believe you but sometimes this even trips me up.
Try making sure that snapshots aren't being re downloaded in the background by intellij. Sometimes that does happen and actually means you have to reinstall.

Sometimes intellij also doesn't pick up the right dependencies and the cache gets in the way. In that case, make sure to delete:
~/.m2/repository/org/deeplearning4j/deeplearning4j-modelimport/1.0.0-SNAPSHOT
~/.m2/repository/org/deeplearning4j/deeplearning4j-nn/1.0.0-SNAPSHOT

This will delete the 2 folders with  the dependencies. Click in to intellij for a second, force it to reload.

Afterwards, re install the dependencies again.

I would also suggest ensure your local check out is up to date with git pull.
Sorry for the busy work, but like I said this trips me up every day and I usually double check my changes propagated.
I'm off for the evening (based in Asia) but I'll check back if you can reply 1 or 2 more times after you double checked all this.

Without being able to reproduce the issue, it will be hard to push any fixes though.

**Comment by joel-a** [comment]
@agibsonccc Thanks for the prompt response.  To your first point: I don't completely understand, are you importing the .h5 file I provided or creating the model with Python code and saving via tf keras?  I'm using Keras 2.2.4

I'm using Netbeans, not Intellij, but I think the same logic applies. I will try that and get back to you. In any case, do you know if this fix will be merged soon? Thank you!

**Comment by joel-a** [comment]
@agibsonccc I think I figured out where the difference between your test and my real use case lies: the shape of the input array. I used the same GAPError.h5 model and passed it an array created with:
```
INDArray input = Nd4j.ones(1,400,128);
```
I got no error; just like you. Now, if I change the shape of the input to (2,400,128):
```
INDArray input = Nd4j.ones(2,400,128);
```
I get the error I've been seeing:
```
Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Arrays are different shape and are not broadcastable. Array 1 shape = [800, 512], array 2 shape = [2, 400, 512]
	at org.nd4j.linalg.api.shape.Shape.assertBroadcastable(Shape.java:3568)
	at org.nd4j.linalg.api.shape.Shape.assertBroadcastable(Shape.java:3561)
	at org.nd4j.linalg.api.shape.Shape.broadcastOutputShape(Shape.java:291)
	at org.deeplearning4j.nn.graph.vertex.impl.ElementWiseVertex.doForward(ElementWiseVertex.java:98)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputOfLayersDetached(ComputationGraph.java:2438)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1740)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1696)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1626)
```

Please let me know if you can replicate the error now by changing the first dimension of the input array to something other than 1.

**Comment by agibsonccc** [comment]
Perfect thanks for the confirmation I will get this worked out tomorrow and
update the test cases with a few batch sizes

**Comment by agibsonccc** [comment]
@joel-a updated my test case to 2 and 3 and seems to work. Batch Normalization required a change as well again.

**Review by treo — APPROVED** [review]

**Comment by joel-a** [comment]
@agibsonccc I'm having issues loading a saved ComputationGraph. I believe it might be related to the latest changes. 

I'm loading an .h5 Keras model using importKerasModelAndWeights (as in previous examples). Then I call save (new File("model.zip)) on the ComputationGraph obj. When I try to load with ComputationGraph.load("model.zip", false), I get the following error:

```
Exception in thread "main" java.lang.RuntimeException: Error deserializing JSON ComputationGraphConfiguration. Saved model JSON is not a valid ComputationGraphConfiguration
	at org.deeplearning4j.util.ModelSerializer.restoreComputationGraphHelper(ModelSerializer.java:560)
	at 
[mask_proxy.zip](https://github.com/eclipse/deeplearning4j/files/6744702/mask_proxy.zip)
org.deeplearning4j.util.ModelSerializer.restoreComputationGraph(ModelSerializer.java:462)
	at org.deeplearning4j.util.ModelSerializer.restoreComputationGraph(ModelSerializer.java:647)
	at org.deeplearning4j.nn.graph.ComputationGraph.load(ComputationGraph.java:4618)
```

Attached you will find two of the models I'm trying to load.  Thank you for your prompt attention to this!

[decoder_proxy.zip](https://github.com/eclipse/deeplearning4j/files/6744707/decoder_proxy.zip)
[mask_proxy.zip](https://github.com/eclipse/deeplearning4j/files/6744708/mask_proxy.zip)

**Comment by agibsonccc** [comment]
@joel-a  this is not a bug. 'edu.mit.ll.seamnet.layers.DownSample0' <- this appears to be a custom layer. Please ensure this is on your classpath when attempting to load the model. If you want help with custom layers, do you mind posting over on the forums with something a bit more fleshed out?
